flink-table/flink-sql-parser-hive/src/main/codegen/includes/parserImpls.ftl

```diff
@@ -218,4 +218,543 @@ SqlNode TableOption() :
+    SqlIdentifier name;
+    SqlDataTypeSpec type;
+    SqlCharStringLiteral comment = null;
+}
+{
+    name = SimpleIdentifier()
+    type = ExtendedDataType()
+    [ <COMMENT> <QUOTED_STRING> {
+        comment = createStringLiteral(token.image, getPos());
+    }]
+    {
+        SqlTableColumn tableColumn = new SqlTableColumn(name, type, null, comment, getPos());
+        list.add(tableColumn);
+    }
+}
+
+void PartColumnDef(List<SqlNode> list) :
+{
+    SqlParserPos pos;
+    SqlIdentifier name;
+    SqlDataTypeSpec type;
+    SqlCharStringLiteral comment = null;
+}
+{
+    name = SimpleIdentifier()
+    type = DataType()
+    [ <COMMENT> <QUOTED_STRING> {
+        comment = createStringLiteral(token.image, getPos());
+    }]
+    {
+        type = type.withNullable(true);
+        SqlTableColumn tableColumn = new SqlTableColumn(name, type, null, comment, getPos());
+        list.add(tableColumn);
+    }
+}
+
+void TableColumn(HiveTableCreationContext context) :
+{
+}
+{
+    (LOOKAHEAD(2)
+        TableColumnWithConstraint(context)
+    |
+        TableConstraint(context)
+    )
+}
+
+/** Parses a table constraint for CREATE TABLE. */
+void TableConstraint(HiveTableCreationContext context) :
+{
+  SqlIdentifier constraintName = null;
+  final SqlLiteral spec;
+  final SqlNodeList columns;
+}
+{
+  [ constraintName = ConstraintName() ]
+  spec = TableConstraintSpec()
+  columns = ParenthesizedSimpleIdentifierList()
+  context.pkTrait = ConstraintTrait()
+  {
+      SqlTableConstraint tableConstraint = new SqlTableConstraint(
+                                            constraintName,
+                                            spec,
+                                            columns,
+                                            SqlConstraintEnforcement.NOT_ENFORCED.symbol(getPos()),
+                                            true,
+                                            getPos());
+      context.constraints.add(tableConstraint);
+  }
+}
+
+SqlLiteral TableConstraintSpec() :
+{
+    SqlLiteral spec;
+}
+{
+    <PRIMARY> <KEY>
+    {
+        spec = SqlUniqueSpec.PRIMARY_KEY.symbol(getPos());
+        return spec;
+    }
+}
+
+SqlIdentifier ConstraintName() :
+{
+    SqlIdentifier constraintName;
+}
+{
+    <CONSTRAINT> constraintName = SimpleIdentifier() {
+        return constraintName;
+    }
+}
+
+void TableColumnWithConstraint(HiveTableCreationContext context) :
+{
+    SqlParserPos pos;
+    SqlIdentifier name;
+    SqlDataTypeSpec type;
+    SqlCharStringLiteral comment = null;
+}
+{
+    name = SimpleIdentifier()
+    type = ExtendedDataType()
+    {
+        // we have NOT NULL column constraint here
+        if (!type.getNullable()) {
+            if (context.notNullTraits == null) {
+                context.notNullTraits = new ArrayList();
+                context.notNullCols = new ArrayList();
+            }
+            context.notNullTraits.add(ConstraintTrait());
+            context.notNullCols.add(name);
```

Comment:
The `ConstraintTrait()` should be optional ?

Reply:
No, each constraint must have a trait. If user doesn't specify the trait we'll return a default value.

Reply:
Fine, usually write the parse block first then the java code, can we follow that pattern ?

Reply:
OK